### DISCUSSION TITLE
Image class and style delay issue

### DISCUSSION THREAD

**Discussion by levipadre** [vercel/next.js#58876] [Help]
Has anyone been having issues with images?
Whenever I load a nextjs Image, and I have an object-fit on the image for example, I always notice a flick when it loads.
I tried defining a class in the main style, like `.img-cover{object-fit: cover}` or adding style to the image, like `style={{objectFit: 'cover'}}`, but on page load I always see first the image without it, so it squeezed.
I made a video with no throttling:
https://app.screencast.com/q5qf2V05QGWjD

I'm loading the component (which contains the image) from page.tsx:
```
import SliderLoading from '@/components/content-blocks/slider/slider-loading';
const Slider = dynamic(() => import('@/components/content-blocks/slider/slider'), {
    loading: () => <SliderLoading />,
    ssr: false
});
```

and the image itself:
```
               <Image
                    alt={image.name }
                    src={image.url}
                    className='img-fluid img-cover'
                    quality={70}
                    width={345}
                    height={345}
                    sizes='(min-width: 1600px) 345px, (min-width: 1400px) 300px, (min-width: 1200px) 255px, (min-width: 1000px) 210px, (min-width: 580px) 510px, 92.31vw'
                    loading='lazy'
                    placeholder='blur'
                    style={{objectFit: 'cover'}}
                    blurDataURL={blurDataUrl}
                />
```

Any ideas or suggestions?

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:27 PDT 2023; root:xnu-10002.41.9~6/RELEASE_X86_64
    Binaries:
      Node: 18.17.0
      npm: 9.6.7
      Yarn: 1.22.19
      pnpm: N/A
    Relevant Packages:
      next: 13.5.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.2.2
    Next.js Config:
      output: N/A
```

**Comment by tollanes** [comment]
I think they're not applying styling to the blur image. Which makes any image that is dynamically sized by CSS basically behave like your example. I really dont understand why this is the case. We are seeing the same issue on our blog hero images which do the weird pop in.. 

Almost tempted to just write my own raw image component for this since the feature seems incomplete. Is this intentional to save performance or something?

**Comment by joepetrillo** [comment]
Running into this issue as well. You can see it happening here in my app quite noticeably.

Stretched
<img width="400" alt="1" src="https://github.com/user-attachments/assets/9d484896-8e19-4781-af42-dc923072397a">

Properly fit
<img width="400" alt="2" src="https://github.com/user-attachments/assets/64facc84-3fd1-4761-b338-9b7d0e0145e4">

Video of it happening

https://github.com/user-attachments/assets/2cf607f3-e6b3-4043-b6f1-1b0602ee5004

**Comment by rikas** [comment]
I also struggle with this issue, remote images with `object-fit: cover` will stretch on page refresh... I can't find a fix for this!

**Comment by sminha** [comment]
I’ve been struggling with this squeezed image flicker for a while and wanted to share my workaround.

As I understand it, the root cause is that object-fit is a layout-dependent property that requires the image's intrinsic aspect ratio to function. Unlike basic dimensions, this ratio is **only available after the browser starts decoding the image data**. Because browsers are optimized to paint pixels as soon as they are decoded to boost performance, they often **render the raw image according to the container's `width/height` before the `object-fit` calculation is complete**. This creates a computational gap where the image appears squeezed for a split second until its actual proportions are processed and applied.

To fix this, I decided to keep the image hidden using `opacity-0` and only reveal it once it's 100% ready. I used the `decode()` method inside `onLoad` to ensure the browser has finished both downloading and decoding the image with the correct styles before switching the opacity.

```tsx
const [isLoaded, setIsLoaded] = useState(false);

<Image
  src={src}
  fill
  className={cn(isLoaded ? 'opacity-100' : 'opacity-0', 'object-cover')}
  onLoad={(e) => {
    const target = e.target as HTMLImageElement;
    target.decode().then(() => setIsLoaded(true));
  }}
/>
```

I know this might make the image appear slightly later, but I personally think making the user wait a tiny bit for a clean fade-in is much better than showing a glitchy, flickering image. While this might not be a perfect fundamental fix, it’s a very effective way to prioritize visual stability and eliminate that annoying jank!

I hope my comments and this approach are helpful to you! 🙂
